helios-testing/src/test/java/com/spotify/helios/testing/DefaultUndeployerTest.java

```diff
@@ -0,0 +1,131 @@
+
+  @Before
+  public void before() {
+    client = mock(HeliosClient.class);
+  }
+
+  @Test
+  public void testSuccess() throws Exception {
+    final ListenableFuture<JobUndeployResponse> undeployFuture1 = Futures
+        .immediateFuture(new JobUndeployResponse(JobUndeployResponse.Status.OK, HOST1, JOB_ID));
+    final ListenableFuture<JobUndeployResponse> undeployFuture2 = Futures
+        .immediateFuture(new JobUndeployResponse(JobUndeployResponse.Status.OK, HOST2, JOB_ID));
+
+    //noinspection unchecked
+    when(client.undeploy(jobIdWithNameAndVersion(JOB_ID), anyString())).thenReturn(
+        undeployFuture1, undeployFuture2);
+
+    when(client.deleteJob(jobIdWithNameAndVersion(JOB_ID))).thenReturn(Futures.immediateFuture(
+        new JobDeleteResponse(JobDeleteResponse.Status.OK)));
+
+    final List<AssertionError> errors = new DefaultUndeployer(client).undeploy(JOB, HOSTS);
+
+    assertThat(errors, Matchers.empty());
+  }
+
+  @Test
+  public void testUndeployError() throws Exception {
+    final ListenableFuture<JobUndeployResponse> undeployFuture1 = Futures.immediateFuture(
+        new JobUndeployResponse(JobUndeployResponse.Status.FORBIDDEN, HOST1, JOB_ID));
+    final ListenableFuture<JobUndeployResponse> undeployFuture2 = Futures.immediateFuture(
+        new JobUndeployResponse(JobUndeployResponse.Status.HOST_NOT_FOUND, HOST2, JOB_ID));
+
+    //noinspection unchecked
+    when(client.undeploy(jobIdWithNameAndVersion(JOB_ID), anyString())).thenReturn(
+        undeployFuture1, undeployFuture2);
+
+    when(client.deleteJob(jobIdWithNameAndVersion(JOB_ID))).thenReturn(Futures.immediateFuture(
+        new JobDeleteResponse(JobDeleteResponse.Status.OK)));
+
+    final List<AssertionError> errors = new DefaultUndeployer(client).undeploy(JOB, HOSTS);
+
+    assertThat(errors.size(), equalTo(2));
+  }
+
+  @Test
+  public void testDeleteError() throws Exception {
+    final ListenableFuture<JobUndeployResponse> undeployFuture1 = Futures.immediateFuture(
+        new JobUndeployResponse(JobUndeployResponse.Status.OK, HOST1, JOB_ID));
+    final ListenableFuture<JobUndeployResponse> undeployFuture2 = Futures.immediateFuture(
+        new JobUndeployResponse(JobUndeployResponse.Status.OK, HOST2, JOB_ID));
+
+    //noinspection unchecked
+    when(client.undeploy(jobIdWithNameAndVersion(JOB_ID), anyString())).thenReturn(
+        undeployFuture1, undeployFuture2);
+
+    when(client.deleteJob(jobIdWithNameAndVersion(JOB_ID))).thenReturn(Futures.immediateFuture(
+        new JobDeleteResponse(JobDeleteResponse.Status.STILL_IN_USE)));
+
+    final List<AssertionError> errors = new DefaultUndeployer(client).undeploy(JOB, HOSTS);
+
+    assertThat(errors.size(), equalTo(1));
+  }
+
+  private static JobId jobIdWithNameAndVersion(final JobId jobId) {
+    return argThat(new ArgumentMatcher<JobId>() {
+      @Override
+      public boolean matches(Object argument) {
+        if (argument instanceof JobId) {
```

Comment:
you can extend `org.hamcrest.CustomTypeSafeMatcher` instead and it takes care of these instanceof checks for you